interface/src/app/analyze/analysis/sampleBin.js

```diff
@@ -13,68 +13,58 @@ angular.module('adage.analyze.sampleBin', [
   return $resource('/api/v0/activity/');
 }])
 
-.factory('SampleBin', ['$log', 'Sample', 'Activity',
-function($log, Sample, Activity) {
+.factory('SampleBin', ['$log', '$cacheFactory', '$q', 'Sample', 'Activity',
+function($log, $cacheFactory, $q, Sample, Activity) {
   var SampleBin = {
     samples: [],
     heatmapData: {},
     sampleData: {},
+    sampleCache: $cacheFactory('sample'),
+    activityCache: $cacheFactory('activity'),
 
-    add_sample: function(id) {
+    addSample: function(id) {
       if (SampleBin.samples.indexOf(+id) !== -1) {
         // quietly ignore the double-add
-        $log.warn('SampleBin.add_sample: ' + id +
+        $log.warn('SampleBin.addSample: ' + id +
             ' already in the sample list; ignoring.');
       } else {
         SampleBin.samples.push(+id);
+        // TODO when cache generalized: start pre-fetching sample data here
       }
     },
 
-    remove_sample: function(id) {
+    removeSample: function(id) {
       var pos = SampleBin.samples.indexOf(+id);
       SampleBin.samples.splice(pos, 1);
-      // TODO need to report query progress and errors somehow
-      Activity.get({sample__in: SampleBin.samples.join()},
-        // success callback
-        function(responseObject, responseHeaders) {
-          if (responseObject) {
-            SampleBin.heatmapData.activity = responseObject.objects;
-            // TODO need to find & report (list) samples that return no results
-          }
-        },
-        // failure callback
-        function(responseObject, responseHeaders) {
-          $log.error('Query errored with: ' + responseObject);
-        }
-      );
+      SampleBin.rebuildHeatmapActivity(SampleBin.samples);
     },
 
-    add_experiment: function(sample_id_list) {
-      for (var i = 0; i < sample_id_list.length; i++) {
-        SampleBin.add_sample(sample_id_list[i]);
+    addExperiment: function(sampleIdList) {
+      for (var i = 0; i < sampleIdList.length; i++) {
+        SampleBin.addSample(sampleIdList[i]);
       }
     },
 
-    add_item: function(search_item) {
-      if (search_item.item_type === 'sample') {
-        SampleBin.add_sample(search_item.pk);
-      } else if (search_item.item_type === 'experiment') {
-        SampleBin.add_experiment(search_item.related_items);
+    addItem: function(searchItem) {
+      if (searchItem.item_type === 'sample') {
+        SampleBin.addSample(searchItem.pk);
+      } else if (searchItem.item_type === 'experiment') {
+        SampleBin.addExperiment(searchItem.related_items);
       }
     },
 
-    has_item: function(search_item) {
-      if (search_item.item_type === 'sample') {
-        if (SampleBin.samples.indexOf(+search_item.pk) !== -1) {
+    hasItem: function(searchItem) {
+      if (searchItem.item_type === 'sample') {
+        if (SampleBin.samples.indexOf(+searchItem.pk) !== -1) {
```

Comment:
Right. It's here as a succinct way to make sure it's an `int`. There is a possibility that `searchItem.pk` winds up a `str` -- I think I saw that happen in debugging, actually.